File: fundamentos/desafio-js/src/questions/top-3-lowest-balance-at-ag-47.ts
import { IAccount } from "../i-account";

/**
 * Você deve mostrar os nomes dos três clientes com menor saldo da agência 47, 
 * separados por vírgula e em ordem crescente (do menor para o maior). 
 * 
 * Qual seria a sua saída do programa?
 */
export default function top3LowestBalanceAtAg47(...accounts: Array<IAccount>) {
  const names = accounts
    .filter(acc => acc.agencia == 47)
    .sort((a, b) => a.balance - b.balance)
    .filter((acc, idx) => idx < 3)
    .map(acc => acc.name)
    .sort();

  console.log(`Os três clientes com menor saldo da ag 47 são: ${names.join(', ')}`);
}

File: fundamentos/desafio-dart/lib/questions/top_3_lowest_balance_at_ag_47.dart
import 'package:desafio_dart/domains/account.dart';

/// Você deve mostrar os nomes dos três clientes com menor saldo da agência 47,
/// separados por vírgula e em ordem crescente (do menor para o maior).
///
/// Qual seria a sua saída do programa?
void top3LowestBalanceAtAg47(List<Account> accounts) {
  final names = (accounts
          .where((acc) => acc.agencia == 47)
          .toList() // Convert iterable to List
            ..sort((a, b) => a.balance.compareTo(b.balance)) // Chaining sort
      ) // return toList (before ..)
      .take(3)
      .map((acc) => acc.name)
      .toList() // Convert Iterable to List
        ..sort(); // Chaining Sort

  print('Os três clientes com menor saldo da ag 47 são: ${names.join(', ')}');
}
